File: backend/src/api/routes/blocks.py
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, status, Query

from ...core.database import get_db
from ...api.dependencies import get_current_admin, validate_temp_link
from ...schemas.block import BlockCreate, BlockUpdate, BlockResponse, BlockReorderRequest
from ...services import block_service

router = APIRouter(prefix="/blocks", tags=["blocks"])


# Admin routes (authenticated)
@router.get("/consultant/{consultant_id}", response_model=list[BlockResponse])
async def get_consultant_blocks(
    consultant_id: int,
    block_type: Literal["project", "skill", "misc", "certification"] | None = Query(default=None),
    _admin: dict = Depends(get_current_admin),
):
    """Get all blocks for a consultant (admin access)"""
    with get_db() as conn:
        blocks = await block_service.get_consultant_blocks(conn, consultant_id, block_type)
    return blocks


# Temporary link routes (no auth, token in URL)
@router.get("/edit/{token}", response_model=list[BlockResponse])
async def get_blocks_via_token(
    token: str,
    block_type: Literal["project", "skill", "misc", "certification"] | None = Query(default=None),
):
    """Get consultant blocks via temporary link"""
    link = await validate_temp_link(token)
    with get_db() as conn:
        blocks = await block_service.get_consultant_blocks(conn, link['consultant_id'], block_type)
    return blocks


@router.post("/edit/{token}", response_model=BlockResponse, status_code=status.HTTP_201_CREATED)
async def create_block_via_token(
    token: str,
    block_data: BlockCreate,
):
    """Create block via temp link"""
    link = await validate_temp_link(token)
    with get_db() as conn:
        block = await block_service.create_block(conn, link['consultant_id'], block_data)
    return block


@router.put("/edit/{token}/{block_id}", response_model=BlockResponse)
async def update_block_via_token(
    token: str,
    block_id: int,
    block_data: BlockUpdate,
):
    """Update block via temp link"""
    link = await validate_temp_link(token)

    # Verify block belongs to consultant
    with get_db() as conn:
        block = await block_service.get_block(conn, block_id)
        if not block or block["consultant_id"] != link["consultant_id"]:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Block not found")

        updated_block = await block_service.update_block(conn, block_id, block_data)
    return updated_block


@router.delete("/edit/{token}/{block_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_block_via_token(
    token: str,
    block_id: int,
):
    """Delete block via temp link"""
    link = await validate_temp_link(token)

    # Verify block belongs to consultant
    with get_db() as conn:
        block = await block_service.get_block(conn, block_id)
        if not block or block["consultant_id"] != link["consultant_id"]:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Block not found")

        await block_service.delete_block(conn, block_id)
    return None


@router.post("/edit/{token}/reorder", status_code=status.HTTP_204_NO_CONTENT)
async def reorder_blocks_via_token(
    token: str,
    reorder_data: BlockReorderRequest,
):
    """Reorder blocks via temp link"""
    link = await validate_temp_link(token)
    with get_db() as conn:
        await block_service.reorder_blocks(conn, link['consultant_id'], reorder_data.block_orders)
    return None

File: frontend/src/stores/blocks.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import api from '@/services/api'

export const useBlocksStore = defineStore('blocks', () => {
  const blocks = ref([])

  const blocksByType = computed(() => {
    const grouped = {
      project: [],
      skill: [],
      misc: [],
      certification: []
    }

    for (const block of blocks.value) {
      if (grouped[block.block_type]) {
        grouped[block.block_type].push(block)
      }
    }

    return grouped
  })

  async function fetchBlocks(consultantId, viaToken = null) {
    try {
      const url = viaToken ? `/blocks/edit/${viaToken}` : `/blocks/consultant/${consultantId}`
      const response = await api.get(url)
      blocks.value = response.data
      return blocks.value
    } catch (error) {
      console.error('Error fetching blocks:', error)
      throw error
    }
  }

  async function createBlock(blockData, viaToken = null) {
    try {
      const url = viaToken ? `/blocks/edit/${viaToken}` : `/blocks`
      const response = await api.post(url, blockData)
      blocks.value.push(response.data)
      return response.data
    } catch (error) {
      console.error('Error creating block:', error)
      throw error
    }
  }

  async function updateBlock(blockId, blockData, viaToken = null) {
    try {
      const url = viaToken ? `/blocks/edit/${viaToken}/${blockId}` : `/blocks/${blockId}`
      const response = await api.put(url, blockData)
      const index = blocks.value.findIndex((b) => b.id === blockId)
      if (index !== -1) {
        blocks.value[index] = response.data
      }
      return response.data
    } catch (error) {
      console.error('Error updating block:', error)
      throw error
    }
  }

  async function deleteBlock(blockId, viaToken = null) {
    try {
      const url = viaToken ? `/blocks/edit/${viaToken}/${blockId}` : `/blocks/${blockId}`
      await api.delete(url)
      blocks.value = blocks.value.filter((b) => b.id !== blockId)
    } catch (error) {
      console.error('Error deleting block:', error)
      throw error
    }
  }

  async function reorderBlocks(blockOrders, viaToken = null) {
    try {
      const url = viaToken ? `/blocks/edit/${viaToken}/reorder` : `/blocks/reorder`
      await api.post(url, { block_orders: blockOrders })
    } catch (error) {
      console.error('Error reordering blocks:', error)
      throw error
    }
  }

  return {
    blocks,
    blocksByType,
    fetchBlocks,
    createBlock,
    updateBlock,
    deleteBlock,
    reorderBlocks
  }
})
